File: smalivm/src/main/java/org/cf/smalivm/opcode/InstanceOfOp.java
package org.cf.smalivm.opcode;

import org.cf.smalivm.VirtualMachine;
import org.cf.smalivm.context.ExecutionNode;
import org.cf.smalivm.context.HeapItem;
import org.cf.smalivm.context.MethodState;
import org.cf.smalivm.dex.CommonTypes;
import org.cf.smalivm.type.ClassManager;
import org.cf.smalivm.type.VirtualType;
import org.cf.util.Utils;
import org.jf.dexlib2.builder.MethodLocation;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class InstanceOfOp extends MethodStateOp {

    private static final Logger log = LoggerFactory.getLogger(InstanceOfOp.class.getSimpleName());

    private final VirtualType referenceType;
    private final int destRegister;
    private final int arg1Register;
    private final VirtualMachine vm;

    InstanceOfOp(MethodLocation location, MethodLocation child, int destRegister, int arg1Register, VirtualType
            referenceType, VirtualMachine vm) {
        super(location, child);
        this.destRegister = destRegister;
        this.arg1Register = arg1Register;
        this.referenceType = referenceType;
        this.vm = vm;
    }

    @Override
    public void execute(ExecutionNode node, MethodState mState) {
        HeapItem item = mState.readRegister(arg1Register);
        Object value = item.getValue();

        if (value == null) {
            mState.assignRegister(destRegister, new HeapItem(false, CommonTypes.BOOL));
            return;
        }

        ClassManager classManager = vm.getClassManager();
        for (String typeName : Utils.getDeclaredAndValueTypeNames(item)) {
            VirtualType itemType = classManager.getVirtualType(typeName);
            if (itemType.instanceOf(referenceType)) {
                mState.assignRegister(destRegister, new HeapItem(true, CommonTypes.BOOL));
                return;
            }
        }

        if (item.isUnknown()) {
            // Since value is unknown, we can't get the actual type of the value. Declared type isn't enough.
            mState.assignRegister(destRegister, HeapItem.newUnknown(CommonTypes.BOOL));
        } else {
            mState.assignRegister(destRegister, new HeapItem(false, CommonTypes.BOOL));
        }
    }

    @Override
    public String toString() {
        return getName() + " r" + destRegister + ", " + arg1Register + ", " + referenceType;
    }

}

File: smalivm/src/main/java/org/cf/smalivm/opcode/InstanceOfOp.kt
package org.cf.smalivm.opcode

import org.cf.smalivm.configuration.Configuration
import org.cf.smalivm.dex.CommonTypes
import org.cf.smalivm.dex.SmaliClassLoader
import org.cf.smalivm.type.ClassManager
import org.cf.smalivm.type.VirtualType
import org.cf.smalivm.ExecutionNode
import org.cf.smalivm.UnresolvedChild
import org.cf.smalivm.Value
import org.jf.dexlib2.builder.MethodLocation
import org.jf.dexlib2.iface.instruction.formats.Instruction22c
import org.jf.dexlib2.iface.reference.TypeReference
import org.slf4j.LoggerFactory

class InstanceOfOp internal constructor(
    location: MethodLocation,
    private val destRegister: Int,
    private val arg1Register: Int,
    private val referenceType: VirtualType,
    private val classManager: ClassManager
) : Op(location) {

    override val registersReadCount = 1
    override val registersAssignedCount = 1

    override fun execute(node: ExecutionNode): Array<out UnresolvedChild> {
        val state = node.state
        val instance = node.state.readRegister(arg1Register)
        if (instance.raw == null) {
            state.assignRegister(destRegister, Value.wrap(false, CommonTypes.BOOLEAN))
            return finishOp()
        }
        for (typeName in instance.declaredAndValueTypeNames) {
            val itemType = classManager.getVirtualType(typeName)
            if (itemType.instanceOf(referenceType)) {
                state.assignRegister(destRegister, Value.wrap(true, CommonTypes.BOOLEAN))
                return finishOp()
            }
        }
        if (instance.isUnknown) {
            // Since value is unknown, we can't get the actual type of the value and declared type isn't specific enough.
            state.assignRegister(destRegister, Value.unknown(CommonTypes.BOOLEAN))
        } else {
            state.assignRegister(destRegister, Value.wrap(false, CommonTypes.BOOLEAN))
        }
        return finishOp()
    }

    override fun toString() = "$name r$destRegister, $arg1Register, $referenceType"

    companion object : OpFactory {
        private val log = LoggerFactory.getLogger(InstanceOfOp::class.java.simpleName)

        override fun build(
            location: MethodLocation,
            classManager: ClassManager,
            classLoader: SmaliClassLoader,
            configuration: Configuration
        ): Op {
            val instr = location.instruction as Instruction22c
            val destRegister = instr.registerA
            val arg1Register = instr.registerB
            val typeReference = instr.reference as TypeReference
            val referenceType = classManager.getVirtualType(typeReference)
            return InstanceOfOp(location, destRegister, arg1Register, referenceType, classManager)
        }
    }
}
